README.md
# sped-gtin

A partir do leyout 4.00 da NFe (e NFCe) o GTIN do produto passa a ser um item OBRIGATÓRIO, qunado existir. Protanto em todas as NFe (e NFCe) que contêm produtos que possuem codigos de barras para fins comerciais (GTIN 8, 12,. 13 ou 14) derão indicar os mesmos em suas notas.

É importantissimo que esses GTIN sejam corretos e validados, antes de serem insereidos em seus sistemas para evitar rejeições por parte da SEFAZ.

Esta classe faz exatamente isso verifica a correção do numero GTIN, com releção a sua estrutura, prefixo, região e digito verificador. A SEFAZ fará cunstultas adicionais ao [Cadastro Nacional de Produtos - CNP](https://www.gs1br.org/servicos-e-solucoes/cadastro-centralizado-de-gtin?gclid=Cj0KCQjw4_zVBRDVARIsAFNI9eCoTYJdZTQ36i4aAWsW4Hmppbqk4BVEvty4gQKXnMnAfX2XRcQcawwaAgv9EALw_wcB). mas ainda não existe a disponibilidade de uma API para que nós também possamos faze-lo, então ainda poderão haver rejeições caso os dados que você possue não estejam de acordo com o CNP. 

[![Build Status][ico-travis]][link-travis]
[![Coverage Status][ico-scrutinizer]][link-scrutinizer]
[![Quality Score][ico-code-quality]][link-code-quality]
[![Code Intelligence Status][ico-code-intelligence]][link-code-intelligence]

[![Latest Stable Version][ico-stable]][link-packagist]
[![Latest Version on Packagist][ico-version]][link-packagist]
[![License][ico-license]][link-packagist]
[![Total Downloads][ico-downloads]][link-downloads]

[![Issues][ico-issues]][link-issues]
[![Forks][ico-forks]][link-forks]
[![Stars][ico-stars]][link-stars]


## Formas de Uso

```php

use NFePHP\Gtin\Gtin;

try {
    $gtin = new Gtin('78935761');
    $region = $gtin->region;
    $prefix = $gtin->prefix;
    $digitoVerificador = $gtin->checkDigit;
    
    if ($gtin->isValid()) {
        echo "Valido";
    }
} catch (\Exception $e) {
    echo $e->getMessage();
}


```
### A classe Gtin, possue várias propriedades 

**region - (string) indica a região de uso do numero ex. GS1 Brasil**

**prefix - (string) indica o prefixo da região ex. 789**

**checkDigit - (int) indica o digito verificador para uma determinada sequência numérica (calculado)**

**restricted - (bool) indica se o pefixo é de uso restrito**

**type -  - (int) indica o tipo de gtin (8,12,13, ou 14)**

### A função principal da classe é :

function isValid()

Essa função não recebe parâmetros e retornará TRUE caso o numero GTIN seja em principio válido.

A classe também pode ser instanciada estaticamente.

```php

use NFePHP\Gtin\Gtin;

try {
    $gtin = Gtin::check('78935761');
    $region = $gtin->region;
    $prefix = $gtin->prefix;
    $digitoVerificador = $gtin->checkDigit;
    
    if ($gtin->isValid()) {
        echo "Valido";
    }
} catch (\Exception $e) {
    echo $e->getMessage();
}
```

## Validações e Exceptions

Caso o numero fornecido não atenda alguma restrição será retornado um Exception

- Um GTIN deve ser passado para a classe.
- GTIN deve conter apenas numeros.
- Apenas GTIN 8, 12, 13 ou 14 esse numero não atende esses parâmetros.
- Um GTIN 14 não pode iniciar com numeral ZERO.
- O prefixo $this->prefix do GTIN é INVALIDO [$this->region].
- O digito verificador é INVALIDO.


Este pacote é aderente com os [PSR-1], [PSR-2] e [PSR-4]. Se você observar negligências de conformidade, por favor envie um patch via pull request.

[PSR-1]: https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-1-basic-coding-standard.md
[PSR-2]: https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-2-coding-style-guide.md
[PSR-4]: https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-4-autoloader.md

Não deixe de se cadastrar no [grupo de discussão do NFePHP](http://groups.google.com/group/nfephp) para acompanhar o desenvolvimento e participar das discussões e tirar dúvidas!

## Install

**Este pacote está listado no [Packgist](https://packagist.org/) foi desenvolvido para uso do [Composer](https://getcomposer.org/), portanto não será explicitada nenhuma alternativa de instalação.**

*E deve ser instalado com:*
```bash
composer require nfephp-org/sped-gtin
```
Ou ainda alterando o composer.json do seu aplicativo inserindo:
```json
"require": {
    "nfephp-org/sped-gtin" : "^1.0"
}
```

*Para utilizar o pacote em desenvolvimento (branch master) deve ser instalado com:*
```bash
composer require nfephp-org/sped-gtin:dev-master
```

*Ou ainda alterando o composer.json do seu aplicativo inserindo:*
```json
"require": {
    "nfephp-org/sped-gtin" : "dev-master"
}
```

> NOTA: Ao utilizar este pacote na versão em desenvolvimento não se esqueça de alterar o composer.json da sua aplicação para aceitar pacotes em desenvolvimento, alterando a propriedade "minimum-stability" de "stable" para "dev".
> ```json
> "minimum-stability": "dev"
> ```


## Requirements

Para que este pacote possa funcionar são necessários os seguintes requisitos do PHP e outros pacotes dos quais esse depende.

- PHP 5.6 ou PHP 7.x (recomendável PHP 7.x) 


## Contributing

Para contribuir com correções de BUGS, melhoria no código, documentação, elaboração de testes ou qualquer outro auxílio técnico e de programação por favor observe o [CONTRIBUTING](CONTRIBUTING.md) e o  [Código de Conduta](CONDUCT.md) para maiores detalhes.

## Change log

Acompanhe o [CHANGELOG](CHANGELOG.md) para maiores informações sobre as alterações recentes.

## Testing

Todos os testes são desenvolvidos para operar com o PHPUNIT

## Security

Caso você encontre algum problema relativo a segurança, por favor envie um email diretamente aos mantenedores do pacote ao invés de abrir um ISSUE.

## Credits

Roberto L. Machado (owner and developer)

## License

Este pacote está diponibilizado sob LGPLv3 ou MIT License (MIT). Leia  [Arquivo de Licença](LICENSE.md) para maiores informações.

[ico-stable]: https://poser.pugx.org/nfephp-org/sped-gtin/version
[ico-stars]: https://img.shields.io/github/stars/nfephp-org/sped-gtin.svg?style=flat-square
[ico-forks]: https://img.shields.io/github/forks/nfephp-org/sped-gtin.svg?style=flat-square
[ico-issues]: https://img.shields.io/github/issues/nfephp-org/sped-gtin.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/nfephp-org/sped-gtin/master.svg?style=flat-square
[ico-scrutinizer]: https://img.shields.io/scrutinizer/coverage/g/nfephp-org/sped-gtin.svg?style=flat-square
[ico-code-quality]: https://img.shields.io/scrutinizer/g/nfephp-org/sped-gtin.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/nfephp-org/sped-gtin.svg?style=flat-square
[ico-version]: https://img.shields.io/packagist/v/nfephp-org/sped-gtin.svg?style=flat-square
[ico-license]: https://poser.pugx.org/nfephp-org/nfephp/license.svg?style=flat-square
[ico-code-intelligence]: https://scrutinizer-ci.com/g/nfephp-org/sped-gtin/badges/code-intelligence.svg?b=master

[link-packagist]: https://packagist.org/packages/nfephp-org/sped-gtin
[link-travis]: https://travis-ci.org/nfephp-org/sped-gtin
[link-scrutinizer]: https://scrutinizer-ci.com/g/nfephp-org/sped-gtin/code-structure
[link-code-quality]: https://scrutinizer-ci.com/g/nfephp-org/sped-gtin
[link-code-intelligence]: https://scrutinizer-ci.com/code-intelligence
[link-downloads]: https://packagist.org/packages/nfephp-org/sped-gtin
[link-author]: https://github.com/nfephp-org
[link-issues]: https://github.com/nfephp-org/sped-gtin/issues
[link-forks]: https://github.com/nfephp-org/sped-gtin/network
[link-stars]: https://github.com/nfephp-org/sped-gtin/stargazers



CHANGELOG.md
# Changelog

All Notable changes to `sped-gtin` will be documented in this file.

Todas as atualizações a partir de 30/03/2018 devem observar os princípios [Mantendo o CHANGELOG](http://keepachangelog.com/).

## 1.0.0-dev 

## Notas da versão:

### Added
- Nothing

### Deprecated
- Nothing

### Fixed
- Nothing

### Removed
- Nothing

### Security
- Nothing


tests/GtinTest.php
<?php

namespace NFePHP\Gtin\Tests;

use NFePHP\Gtin\Gtin;
use PHPUnit\Framework\TestCase;

class GtinTest extends TestCase
{
    /**
     * @covers Gtin::getPrefix
     * @covers Gtin::getType
     * @covers Gtin::getPrefixRegion
     * @covers Gtin::getCheckDigit
     * @covers Gtin::__contruct
     */
    public function testCanInstantiate()
    {
        $gtin = new Gtin('78935761');
        $this->assertInstanceOf('NFePHP\Gtin\Gtin', $gtin);
    }
    
    /**
     * @covers Gtin::getPrefix
     * @covers Gtin::getType
     * @covers Gtin::getPrefixRegion
     * @covers Gtin::getCheckDigit
     * @covers Gtin::__contruct
     * @covers Gtin::check
     */
    public function testCanInstantiateStatic()
    {
        $gtin = Gtin::check('78935761');
        $this->assertInstanceOf('NFePHP\Gtin\Gtin', $gtin);
    }

    /**
     * @covers Gtin::getPrefixRegion
     */
    public function testRegion()
    {
        $gtin = new Gtin('78935761');
        $this->assertEquals('GS1 Brasil', $gtin->region);
        
        $region = Gtin::check('78935761')->region;
        $this->assertEquals('GS1 Brasil', $region);
    }
    
    /**
     * @covers Gtin::getCheckDigit
     */
    public function testCheckDigit()
    {
        $gtin = new Gtin('78935761');
        $this->assertEquals(1, $gtin->checkDigit);
        
        $dv = Gtin::check('78935761')->checkDigit;
        $this->assertEquals(1, $dv);
    }
    
    /**
     * @covers Gtin::getPrefix
     * @covers Gtin::getType
     */
    public function testPrefix()
    {
        $gtin = new Gtin('78935761');
        $this->assertEquals('789', $gtin->prefix);
        
        $prefix = Gtin::check('78935761')->prefix;
        $this->assertEquals('789', $prefix);
    }
    
    public function testIsValid()
    {
        $gtin = new Gtin('78935761');
        $this->assertTrue($gtin->isValid());
        
        $resp = Gtin::check('78935761')->isValid();
        $this->assertTrue($resp);
    }
    
    /**
     *  @expectedException \InvalidArgumentException
     *  @expectedExceptionMessage Um GTIN deve ser passado para a classe.
     */
    public function testFailStringEmptyGtin()
    {
        $gtin = new Gtin('');
    }
    
    /**
     *  @expectedException \InvalidArgumentException
     *  @expectedExceptionMessage Um GTIN deve ser passado para a classe.
     */    
    public function testFailNullGtin()
    {
        $gtin = new Gtin(null);
    }
    
    /**
     * @expectedException \InvalidArgumentException
     * @expectedExceptionMessage GTIN deve conter apenas numeros.
     */
    public function testInvalidStringGtin()
    {
        $gtin = new Gtin('A12345');
    }
    
    /**
     * @expectedException \InvalidArgumentException
     * @expectedExceptionMessage Apenas GTIN 8, 12, 13 ou 14 esse numero não atende esses parâmetros.
     */
    public function testInvalidMinLengthGtin()
    {
        $gtin = new Gtin('12345');
    }

    /**
     * @expectedException \InvalidArgumentException
     * @expectedExceptionMessage Apenas GTIN 8, 12, 13 ou 14 esse numero não atende esses parâmetros.
     */
    public function testInvalidMaxLengthGtin()
    {
        $gtin = new Gtin('1234567890123456');
    }    
    
    /**
     * @expectedException \InvalidArgumentException
     * @expectedExceptionMessage Apenas GTIN 8, 12, 13 ou 14 esse numero não atende esses parâmetros.
     */
    public function testInvalidLengthGtin()
    {
        $gtin = new Gtin('1234567890');
    }
    
    /**
     * @expectedException \InvalidArgumentException
     * @expectedExceptionMessage O prefixo 021 do GTIN é INVALIDO [Números de circulação restrita dentro da região].
     */
    public function testInvalidPrefixGtin()
    {
        $gtin = Gtin::check('0219907267612')->isValid();
    }
    
    /**
     * @expectedException \InvalidArgumentException
     * @expectedExceptionMessage O digito verificador é INVALIDO.
     */
    public function testInvalidCheckDigit()
    {
        $resp = Gtin::check('7890142547851')->isValid();
    }
    
    /**
     * @expectedException \InvalidArgumentException
     * @expectedExceptionMessage Um GTIN 14 não pode iniciar com numeral ZERO.
     */
    public function testFailGtin14BeginsZero()
    {
        $resp = Gtin::check('07890142547852')->isValid();
    }
}

src/Gtin.php
<?php

namespace NFePHP\Gtin;

/**
 * Class for validation of GTIN numbers used in NFe
 *
 * @category  NFePHP
 * @package   NFePHP\Gtin
 * @copyright NFePHP Copyright (c) 2008-2018
 * @license   http://www.gnu.org/licenses/lgpl.txt LGPLv3+
 * @license   https://opensource.org/licenses/MIT MIT
 * @license   http://www.gnu.org/licenses/gpl.txt GPLv3+
 * @author    Roberto L. Machado <linux.rlm at gmail dot com>
 * @link      http://github.com/nfephp-org/sped-nfe for the canonical source repository
 */

class Gtin
{
    /**
     * @var string
     */
    public $prefix;
    /**
     * @var boolean
     */
    public $restricted;
    /**
     * @var string
     */
    public $region;
    /**
     * @var integer
     */
    public $checkDigit;
    /**
     * @var integer
     */
    public $type;
    /**
     * @var string
     */
    protected $number;
    /**
     * @var integer
     */
    protected $lenght;
    /**
     * @var array
     */
    protected $stdPrefixCollection;
    
    /**
     * Class constructor
     * @param string $gtin
     * @throws \InvalidArgumentException
     */
    public function __construct($gtin)
    {
        $this->stdPrefixCollection = json_decode(file_get_contents(__DIR__.'/prefixcollection.json'));
        if (empty($gtin)) {
            throw new \InvalidArgumentException('Um GTIN deve ser passado para a classe.');
        }
        if (preg_match('/[^0-9]/', $gtin)) {
            throw new \InvalidArgumentException('GTIN deve conter apenas numeros.');
        }
        $this->lenght = (int) strlen($gtin);
        if ($this->lenght != 8 && $this->lenght != 12 && $this->lenght != 13 && $this->lenght != 14) {
            throw new \InvalidArgumentException(
                "Apenas GTIN 8, 12, 13 ou 14 esse numero não atende esses parâmetros."
            );
        }
        $this->number = $gtin;
        $this->prefix = $this->getPrefix($gtin);
        $this->region = $this->getPrefixRegion($this->prefix);
        $this->checkDigit = $this->getCheckDigit($gtin);
        $this->type = $this->getType($gtin);
    }
    
    /**
     * Static instantiation off class
     * @param string $gtin
     * @return \NFePHP\Gtin\Gtin
     */
    public static function check($gtin)
    {
        return new static($gtin);
    }
    
    /**
     * Validate GTIN 8, 12, 13, or 14 with check digit
     * @return boolean
     */
    public function isValid()
    {
        if ($this->lenght == 14 && substr($this->number, 0, 1) == '0') {
            //first digit of GTIN14 can not be zero
            throw new \InvalidArgumentException(
                "Um GTIN 14 não pode iniciar com numeral ZERO."
            );
        }
        if (!$this->isPrefixValid()) {
            throw new \InvalidArgumentException(
                "O prefixo $this->prefix do GTIN é INVALIDO [$this->region]."
            );
        }
        $dv = (int) substr($this->number, -1);
        if ($dv !== $this->checkDigit) {
            throw new \InvalidArgumentException(
                "O digito verificador é INVALIDO."
            );
        }
        return true;
    }
    
    /**
     * Extract region prefix
     * @param string $gtin
     * @return string
     */
    protected function getPrefix($gtin)
    {
        $type = $this->getType($gtin);
        switch ($type) {
            case 14: //begins with number not zero
                return substr($gtin, 1, 3);
            default:
                return substr($gtin, 0, 3);
        }
    }
    
    /**
     * Identify GTIN type GTIN 8,12,13,14 or NONE
     * @param string $gtin
     * @return int
     */
    protected function getType($gtin)
    {
        $gtinnorm = str_pad($gtin, 14, '0', STR_PAD_LEFT);
        if (substr($gtinnorm, 0, 6) == '000000') {
            //GTIN 8
            return 8;
        } elseif (substr($gtinnorm, 0, 2) == '00') {
            //GTIN 12
            return 12;
        } elseif (substr($gtinnorm, 0, 1) == '0') {
            //GTIN 13
            return 13;
        } elseif (substr($gtinnorm, 0, 1) != '0') {
            //GTIN 14
            return 14;
        }
        return 0;
    }
    
    /**
     * Validate prefix region
     * @return boolean
     */
    protected function isPrefixValid()
    {
        return !$this->restricted;
    }
    
    /**
     * Recover region from prefix code
     * @param string $prefix
     * @return string
     */
    protected function getPrefixRegion($prefix)
    {
        $pf = (int) $prefix;
        foreach ($this->stdPrefixCollection as $std) {
            $nI = (int) $std->nIni;
            $nF = (int) $std->nFim;
            $this->restricted = (boolean) $std->restricted;
            $region = $std->region;
            if ($pf >= $nI && $pf <= $nF) {
                return $region;
            }
        }
        return "Not Found";
    }
    
    /**
     * Calculate check digit from GTIN 8, 12, 13 or 14
     * @param string $gtin
     * @return integer
     */
    public function getCheckDigit($gtin)
    {
        $len = (int) strlen($gtin);
        $gtin = substr($gtin, 0, $len-1);
        $gtin = str_pad($gtin, '15', '0', STR_PAD_LEFT);
        $total = 0;
        for ($pos=0; $pos<15; $pos++) {
            $total += ((($pos+1) % 2) * 2 + 1) * $gtin[$pos];
        }
        $dv = 10 - ($total % 10);
        return $dv;
    }
}
